core/src/test/java/org/infinispan/stream/DistributedWriteBehindStreamIteratorTest.java

```diff
@@ -0,0 +1,129 @@
+import org.testng.annotations.Test;
+
+@Test(groups = {"functional", "smoke"}, testName = "iteration.DistributedWriteBehindStreamIteratorTest")
+public class DistributedWriteBehindStreamIteratorTest extends BaseSetupStreamIteratorTest {
+
+   private boolean asyncStore;
+   private boolean sharedStore;
+
+   public DistributedWriteBehindStreamIteratorTest() {
+      // Cache Mode is provided in factory methods
+      super(false, null);
+   }
+
+   DistributedWriteBehindStreamIteratorTest async(boolean asyncStore) {
+      this.asyncStore = asyncStore;
+      return this;
+   }
+
+   DistributedWriteBehindStreamIteratorTest shared(boolean sharedStore) {
+      this.sharedStore = sharedStore;
+      return this;
+   }
+
+   @Override
+   public Object[] factory() {
+      return new Object[]{
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(true).cacheMode(CacheMode.REPL_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(true).cacheMode(CacheMode.REPL_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(false).cacheMode(CacheMode.REPL_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(false).cacheMode(CacheMode.REPL_SYNC),
+
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(false).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(false).cacheMode(CacheMode.DIST_SYNC),
+      };
+   }
+
+   @Override
+   protected Object[] parameterValues() {
+      return concat(super.parameterValues(), asyncStore, sharedStore);
+   }
+
+   @Override
+   protected String[] parameterNames() {
+      return concat(super.parameterNames(), "asyncStore", "sharedStore");
+   }
+
+   @Override
+   protected void enhanceConfiguration(ConfigurationBuilder builder) {
+      DummyInMemoryStoreConfigurationBuilder dimscb = builder.persistence().addStore(DummyInMemoryStoreConfigurationBuilder.class);
+      if (sharedStore) {
+         dimscb.shared(true);
+      }
+      if (asyncStore) {
+         dimscb.storeName(getTestName())
+               .async().enable();
+      }
+   }
+
+   @Test
+   public void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStoreWithReHash() throws Exception {
+      testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(true);
+   }
+
+   @Test
+   public void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStoreWithoutRehash() throws Exception {
+      testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(false);
+   }
+
+   protected void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(boolean rehashAware) {
```

Comment:
I don't like the test name, because the test asserts that a stream request **is** sent to other nodes in some configurations. I'd call it something like `testBackupSegmentsOptimizationWithWriteBehindStore` and get `rehashAware` from a `DataProvider`.

Reply:
Okay